proxy/agent_test.go

```diff
@@ -0,0 +1,226 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy
+
+import (
+	"fmt"
+	"testing"
+	"time"
+)
+
+func TestStartStop(t *testing.T) {
+	current := -1
+	stop := make(chan struct{})
+	desired := "config"
+	start := func(config interface{}, epoch int) error {
+		if current != -1 {
+			t.Error("Expected epoch not to be set")
+		}
+		if epoch != 0 {
+			t.Error("Expected initial epoch to be 0")
+		}
+		if config != desired {
+			t.Errorf("Start got config %v, want %v", config, desired)
+		}
+		current = epoch
+		return nil
+	}
+	cleanup := func(epoch int) {
+		if current != 0 {
+			t.Error("Expected epoch to be set")
+		}
+		if epoch != 0 {
+			t.Error("Expected initial epoch in cleanup to be 0")
+		}
+		close(stop)
+	}
+	a := NewAgent(start, cleanup, 10, time.Millisecond)
+	go a.Run(stop)
+	a.Apply(desired)
+	<-stop
+}
+
+func TestApplyTwice(t *testing.T) {
+	stop := make(chan struct{})
+	desired := "config"
+	start := func(config interface{}, epoch int) error {
+		if epoch == 1 {
+			t.Error("Should start only once for same config")
+		}
+		<-stop
+		return nil
+	}
+	cleanup := func(epoch int) {}
+	a := NewAgent(start, cleanup, 10, time.Millisecond)
+	go a.Run(stop)
+	a.Apply(desired)
+	a.Apply(desired)
+	close(stop)
+}
+
+// TestApplyThrice checks that the current config prevents restart even if the bad config was supplied
+// in between.
+func TestApplyThrice(t *testing.T) {
+	stop := make(chan struct{})
+	good := "good"
+	bad := "bad"
+	applied := false
+	var a Agent
+	start := func(config interface{}, epoch int) error {
+		if config == bad {
+			return fmt.Errorf(bad)
```

Comment:
nit: `return errors.New(bad)` here and elsewhere where string formatting is not needed.